site/app/controllers/admin/ReportController.php

```diff
@@ -776,8 +775,55 @@ public function displayGradebook() {
             new WebResponse(
                 ['admin', 'Report'],
                 'showFullGradebook',
-                $grade_file
+                $grade_file,
+                $grade_summaries_last_run
             )
         );
     }
+
+    /**
+     * Generate a custom filename for the downloaded CSV file
+     */
+    private function generateCustomFilename(): string {
+        $course = $this->core->getConfig()->getCourse();
+        $timestamp = DateUtils::getFileNameTimeStamp();
+        return "{$course}_rainbow_grades_{$timestamp}.csv";
+    }
+
+
+    /**
+     * Download CSV file for Rainbow Grades
+     */
+    #[Route("/courses/{_semester}/{_course}/reports/rainbow_grades_csv")]
+    public function downloadRainbowGradesCSVFile(): void {
+        // Path to the CSV file for Rainbow Grades
+        $csvFilePath = FileUtils::joinPaths(
+            '/var/local/submitty/courses',
+            $this->core->getConfig()->getTerm(),
+            $this->core->getConfig()->getCourse(),
+            'rainbow_grades',
+            'output.csv'
+        );
+
+
+        // Check if the file exists
+        if (file_exists($csvFilePath)) {
+            // Set appropriate headers for file download with a custom filename
+            header('Content-Description: File Transfer');
+            header('Content-Type: application/csv');
+            header('Content-Disposition: attachment; filename="' . $this->generateCustomFilename() . '"');
+            header('Expires: 0');
+            header('Cache-Control: must-revalidate');
+            header('Pragma: public');
+            header('Content-Length: ' . filesize($csvFilePath));
```

Comment:
Could this use the DownloadResponse from 'app\libraries\response\DownloadResponse' instead of setting headers manually?

Reply:
If you to file_get_contents($csvFilePath) instead of readfile() does it work?

Reply:
Unfortunately no, it just returns an empty page.

Reply:
You have to return the download response for it to work.

Reply:
If I do that PHP gets mad and says `Fatal error: A void function must not return a value in /usr/local/submitty/site/app/controllers/admin/ReportController.php on line 826`

Reply:
Fixed in 23c6f701bcd9d1650355e79cdd1006c539af06bc